### ISSUE TITLE
join/intersperse method

### ISSUE THREAD

**Issue by nafg** [scala/collection-strawman#208]
Right now collections have a `.mkString` method, but it's really just a degenerate case of a more useful and general operation, coupled together with some other operations.

`xs.mkString(", ")` can be seen as doing three things:
1. `val ys: C[String] = xs.map(_.toString)` -- _This operation already exists_
2. Interpolate `", "`, a `String`, between the existing `String` values -- _There is currently no way to do this alone!_
3. `Seq[String].flatten` yields a `Seq[Char]` so it's not exactly that, but it's basically doing a flatten with the result collection type of `String`. -- Not sure if there is a way to do this but there should be (well the problem is really only with String and for that there is the no-arg `.mkString`...)

It's nice to have a convenience method like `.mkString` but we should be able to interpolate with other types than String, and without flattening the result.

**Comment by julienrf** [comment]
I definitely agree that `interpolate` would be super useful. I would name it `intersperse` like in [akka-streams](http://doc.akka.io/api/akka/current/akka/stream/scaladsl/Source.html#intersperse[T%3E:Out](start:T,inject:T,end:T):FlowOps.this.Repr[T]), though.

**Comment by hrhino** [comment]
`intersperse` is also [consistent with Haskell](http://hackage.haskell.org/package/base-4.10.0.0/docs/Data-List.html#v:intersperse), too.

**Comment by nafg** [comment]
Agreed. I think it's what I meant to write so I edited the issue summary.

**Comment by Ichoran** [comment]
What is the type signature supposed to be?  `def intersperse[B >: A](in: Iterable[B]): CC[B]`?

**Comment by julienrf** [comment]
I would rather have:

~~~ scala
def intersperse[B >: A](b: B): CC[B]
~~~

**Comment by Ichoran** [comment]
So `mkString(s)` would be equivalent to `.map(_.toString).intersperse(s).flatten`?  Sounds reasonable.

**Comment by nafg** [comment]
@Ichoran yes except that as I pointed out, in fact `.flatten` would produce a `List[Char]` since it uses the outer collection type. At least in the current collections. But it should definitely be equivalent to some combination of smaller more reusable parts. At least the non-nullary versions. `.mkString` without arguments doesn't really have to be equivalent to a combination of other primitives, unless there is something that just does `Seq[String] => String`, then `.mkString` would be a combination of `.map(_.toString)` and that.

**Comment by nafg** [comment]
@julienrf agreed. Although @Ichoran's suggestion could be useful. Maybe it could be called `flatZip` or something. If only there was a straightforward Tuple=>Seq, e.g.:

```scala
// given
def pair2seq[A](t: (A, A)) = Seq(t._1, t._2)
// you could do
(xs zip ys).flatMap(pair2seq)
```

**Comment by Ichoran** [comment]
There is an alternate formulation of `flatten` that could be called `ligate` that flattens by doing a concat of the inner collection type rather than keeping the outer one.  (When talking about DNA, a "ligation" is using an enzyme to glue pieces of DNA together.)  Something like `concatAll` might be more meaningful to programmers, however.  This would also be handy for doing multiple concats; you could just `Array(me, you, them).concatAll` to get the same thing as `me ++ you ++ them`.  (Obviously most useful when the number of items is variable.)

**Comment by EPronovost** [comment]
@julienrf is this what you had in mind for `intersperse`?  Should we override it in other collection types? https://github.com/EPronovost/collection-strawman/commit/8a55f77756042806c27b488908262c77c2b6ab26

**Comment by julienrf** [comment]
You should write a view-based implementation: the underlying `Iterable` should be consumed by `intersperse` only if `fromIterable` does consume the view. I definitely need to address #196 so that external contributions will be easier to make :)

**Comment by ShaneDelmore** [comment]
Would intercalate be appropriate here as well?  http://hackage.haskell.org/package/base-4.10.0.0/docs/Data-List.html#v:intercalate  If not intersperse...flatten is close enough for my needs.

**Comment by EPronovost** [comment]
@julienrf Is this what you mean by a view-based implementation? https://github.com/EPronovost/collection-strawman/commit/33bca5978e2c4ef404d0b00ac42c82d85e118bf9

**Comment by Ichoran** [comment]
@EPronovost - That looks good.

**Comment by EPronovost** [comment]
#224

**Comment by julienrf** [comment]
Closed by #224
